desc/compute/utils.py

```diff
@@ -1349,6 +1350,334 @@ def body(i, mins):
+    resolution : int
+        Number of quadrature points used to compute the bounce integral.
+
+    Returns
+    -------
+    bi : callable
+        This callable method computes the bounce integral F_ℓ(λ) for every
+        specified field line ℓ (constant rho and alpha), for every λ value in
+        ``lambda_pitch``.
+
+    Examples
+    --------
+    .. code-block:: python
+
+        bi = bounce_integral(eq)
+        F = bi(name, lambda_pitch)
+
+    """
+    if rho is None:
+        rho = jnp.linspace(0, 1, 10)
+    if alpha is None:
+        alpha = jnp.linspace(0, (2 - eq.sym) * jnp.pi, 20)
+
+    zeta = np.linspace(0, zeta_max, resolution)
+    grid, data = field_line_to_desc_coords(rho, alpha, zeta, eq)
+    data = eq.compute(
+        ["B^zeta", "|B|", "|B|_z constant rho alpha"], grid=grid, data=data
+    )
+    # TODO: https://github.com/f0uriest/interpax/issues/19
+    coef = CubicHermiteSpline(
+        zeta,
+        data["|B|"].reshape(alpha.size, rho.size, zeta.size),
+        data["|B|_z constant rho alpha"].reshape(alpha.size, rho.size, zeta.size),
+        axis=-1,
+        extrapolate="periodic",
+    ).c
+    der = polyder(coef)
+    # There are zeta.size - 1 splines per field line.
+    assert coef.shape == (4, zeta.size - 1, alpha.size, rho.size)
+    assert der.shape == (3, zeta.size - 1, alpha.size, rho.size)
+
+    def _bounce_integral(name, lambda_pitch):
+        """Compute the bounce integral of the named quantity.
+
+        Parameters
+        ----------
+        name : ndarray
+            Name of quantity in ``data_index`` to compute the bounce integral of.
+        lambda_pitch : ndarray
+            λ values to evaluate the bounce integral at.
+
+        Returns
+        -------
+        F : ndarray, shape(rho.size, alpha.size, resolution, 2, lambda_pitch.size)
+            Bounce integrals evaluated at ``lambda_pitch`` for every field line.
+
+        """
+        # Gauss-Quadrature is expensive to perform because evaluating the integrand
+        # at the optimal quadrature points along the field line would require
+        # root finding to map field line coordinates to desc coordinates.
+        # Newton-Cotes quadrature is inaccurate as the bounce points are not
+        # guaranteed to be near the fixed quadrature points. Instead, we
+        # construct interpolating cubic splines of the integrand and integrate
+        # the spline exactly between the bounce points. This should give
+        # comparable results to a composite Simpson Newton-Cotes with quadrature
+        # points near bounce points.
+        lambda_pitch = jnp.atleast_1d(lambda_pitch)
+        bp = cubic_poly_roots(coef, 1 / lambda_pitch)
+        bp = jnp.moveaxis(bp, 0, -(lambda_pitch.size > 1))
+        assert bp.shape == (zeta.size - 1, alpha.size, rho.size, 3, lambda_pitch.size)
+        b_norm_z = polyeval(der, bp)
+        wells = (b_norm_z[:, :, :, :-1] <= 0) & (b_norm_z[:, :, :, 1:] >= 0)
```

Comment:
Everything above this can be precomputed if `lambda_pitch` is given as an argument to the parent function.